pkg/hubble/metrics/api/context.go

```diff
@@ -418,71 +309,101 @@ func getK8sAppFromLabels(labels []string) string {
+// that leverage the response/return flows where the source and destination are
+// swapped from the request flow.
+func (o *ContextOptions) GetLabelValuesInvertSourceDestination(flow *pb.Flow) (labels []string, err error) {
+	return o.getLabelValues(true, flow)
+}
+
+// getLabelValues returns the values of the context relevant labels according
+// to the configured options. The order of the values is the same as the order
+// of the label names returned by GetLabelNames(). If invert is true, the
+// source and destination related labels are inverted.
+func (o *ContextOptions) getLabelValues(invert bool, flow *pb.Flow) (labels []string, err error) {
 	if len(o.Labels) != 0 {
-		labelsContextLabels, err := labelsContext(o.Labels, flow)
+		labelsContextLabels, err := labelsContext(invert, o.Labels, flow)
 		if err != nil {
 			return nil, err
 		}
 		labels = append(labels, labelsContextLabels...)
 	}
 
-	if len(o.Source) != 0 {
-		var context string
-		for _, source := range o.Source {
-			switch source {
-			case ContextNamespace:
-				context = sourceNamespaceContext(flow)
-			case ContextIdentity:
-				context = sourceIdentityContext(flow)
-			case ContextPod:
-				context = sourcePodContext(flow)
-			case ContextPodShort:
-				context = sourcePodShortContext(flow)
-			case ContextDNS:
-				context = sourceDNSContext(flow)
-			case ContextIP:
-				context = sourceIPContext(flow)
-			case ContextReservedIdentity:
-				context = sourceReservedIdentityContext(flow)
-			}
-			// always use first non-empty context
-			if context != "" {
-				break
-			}
+	var sourceLabel string
+	for _, contextID := range o.Source {
+		sourceLabel = getContextIDLabelValue(contextID, flow, true)
+		// always use first non-empty context
+		if sourceLabel != "" {
+			break
 		}
-		labels = append(labels, context)
 	}
 
-	if len(o.Destination) != 0 {
-		var context string
-		for _, destination := range o.Destination {
-			switch destination {
-			case ContextNamespace:
-				context = destinationNamespaceContext(flow)
-			case ContextIdentity:
-				context = destinationIdentityContext(flow)
-			case ContextPod:
-				context = destinationPodContext(flow)
-			case ContextPodShort:
-				context = destinationPodShortContext(flow)
-			case ContextDNS:
-				context = destinationDNSContext(flow)
-			case ContextIP:
-				context = destinationIPContext(flow)
-			case ContextReservedIdentity:
-				context = destinationReservedIdentityContext(flow)
-			}
-			// always use first non-empty context
-			if context != "" {
-				break
-			}
+	var destinationLabel string
+	for _, contextID := range o.Destination {
+		destinationLabel = getContextIDLabelValue(contextID, flow, false)
+		// always use first non-empty context
+		if destinationLabel != "" {
+			break
 		}
-		labels = append(labels, context)
 	}
 
+	if invert {
+		sourceLabel, destinationLabel = destinationLabel, sourceLabel
+	}
+	if len(o.Source) != 0 {
```

Comment:
Shouldn't the condition rather be

```suggestion
	if sourceLabel != "" {
```

Reply:
sourceLabel will potentially still be `""` when no context options match, even if they're supplied, but we don't want to include the label unless the context option is supplied because we will have a mismatched number of label values and label names, so we have to check `o.Source` (which is only non-empty if the sourceContext is supplied).